numeric_ranges_algorithms/numeric_ranges_algorithms.md

```diff
@@ -0,0 +1,867 @@
+
+Regarding (2), we make the parallel algorithms proposed here take sized random access ranges, as P3179R8 does.  For consistency, we also propose that the output ranges be sized.  As a result, any parallel algorithms with an output range need to return both an iterator to one past the last element of the input, and an iterator to one past the last element of the input.  This tells callers whether there was enough room in the output, and if not, where to start when processing the rest of the input.  This includes all the `*{ex,in}clusive_scan` algorithms we propose.
+
+Difference (3) relates to P3179R8 only proposing parallel algorithms.  It would make sense for us to relax this requirement for the non-parallel algorithms we propose.  This leaves us with two possibilities:
+
+* (single-pass) input and output ranges, the most general; or
+
+* (multipass) forward ranges.
+
+The various reduction and scan algorithms we propose can combine the elements of the range in any order.  For this reason, we make the non-parallel algorithms take (multipass) forward ranges, even though this is not consistent with the existing non-parallel `<numeric>` algorithms.  If users have single-pass iterators, they should just call one of the `fold_*` algorithms, or use the `views::scan` proposed elsewhere.  This has the benefit of letting us specify `ranges::reduce` to return just the value.  We don't propose a separate algorithm `reduce_with_iter`, as we explain elsewhere in this proposal.
+
+## Constexpr parallel algorithms?
+
+<a href="https://wg21.link/p2902r1">P2902R1</a> proposes to add `constexpr` to the parallel algorithms.  P3179R8 does not object to this; see Section 2.10.  We continue the approach of P3179R8 in not opposing P2902R1's approach, but also not depending on it.
+
+## `ranges::reduce` design
+
+In this section, we focus on `ranges::reduce`'s design.  The discussion here applies generally to the other algorithms we propose.
+
+### No default parameters
+
+Section 5.1 of P2760R1 states:
+
+> One thing is clear: `ranges::reduce` should *not* take a default binary operation *nor* a default initial parameter. The user needs to supply both.

```

Comment:
I agree with "no defaults" for `reduce` and `exclusive_scan`, however for `inclusive_scan` I think it is totally fine to omit the initial value parameter, as it naturally comes as the first element of the input.

Reply:
@akukanov That's a good point!  C++17 `inclusive_scan` and `exclusive_scan` already differ.